### DISCUSSION TITLE
I doubt benchmarks be fair !!!

### DISCUSSION THREAD

**Discussion by Rustixir** [ergo-services/ergo#149] [General]
https://erlangforums.com/t/ring-benchmark-erlang-vs-go/684/12

in erlang community we did some benchkmarks around (raw golang) and ( erlang )
erlang was winner in message passing.
erlang internally used lock-free algorithm for registry and  communication between mailbox.
but golang used mutex for channels !!! and in your framework heavily using mutex for synchronization !!!

how is ergo with overhead of framework can beats erlang ??!!!

**Comment by halturin** [comment]
We are moving the same way https://twitter.com/halturin/status/1634165290564034562
![image](https://user-images.githubusercontent.com/118860/227534185-fefc4055-6dbc-434f-8222-b7c84712e4ef.png)

**Comment by halturin** [comment]
>how is ergo with overhead of framework can beats erlang ??!!!

in terms of network messaging Erlang has poor performance due to the ETF encoder/decoder (lists/maps especially. as you may know, strings in Erlang are lists). Besides that,  Ergo ETF has stack-less implementation

>your framework heavily using mutex for synchronization

it depends on where it's used. In Ergo we use most of the time RWMutex, which is not the same as regular Mutex in terms of performance.

**Comment by halturin** [comment]
I move it to the Discussion. Hope you don't mind.

**Comment by halturin** [comment]
In case you wonder about MPSC implementation in Ergo https://github.com/ergo-services/ergo/blob/master/lib/mpsc.go

Here is the benchmark channel vs MPSC queue https://github.com/ergo-services/ergo/blob/master/lib/mpsc_test.go#L188
![image](https://user-images.githubusercontent.com/118860/227538852-db58e1ea-0906-4c93-a917-7d0477556348.png)

**Comment by Rustixir** [comment]
It looks good . 
Thanks for your attention

**Comment by Rustixir** [comment]
But erlang can still beat ergo in many situations !!

For example : 

 When need in-memory transactional systems
  Like RealTime bidding ( erlang have ets, actually is C code and don't make garbage )

**Reply by halturin** [reply]
>But erlang can still beat ergo in many situations !!

not sure why you are arguing this way :)

>When need in-memory transactional systems

there are so many different DBs implemented in Go :)

>erlang have ets, actually is C code and don't make garbage

easy to check with golang's map (which is highly optimized), for example.

so, what is the point of your message?

**Reply by Rustixir** [reply]
Golang map is simple hashmap data structures

Erlang ets use judy array with highly optimize locking ( Judy arrays are highly optimized 256-ary radix tree )


Golang map make garbage and 
have pressure on GC

But 

Erlang ets is C based implemention
and don't make any garbage
